File: 29Items-Projects/7Algorithms/1-Sorting-and-Searching/src/java/com/sortingsearching/sorting/InsertionSort.java
package com.sortingsearching.sorting;

import java.util.Arrays;

public final class InsertionSort {
    private InsertionSort() {
    }

    public static int[] sort(int[] input) {
        int[] values = Arrays.copyOf(input, input.length);
        for (int i = 1; i < values.length; i++) {
            int key = values[i];
            int j = i - 1;
            while (j >= 0 && values[j] > key) {
                values[j + 1] = values[j];
                j--;
            }
            values[j + 1] = key;
        }
        return values;
    }

    public static void main(String[] args) {
        int[] sample = {42, 7, 19, 3, 7, 99, 1};
        System.out.println("InsertionSort");
        System.out.println("input : " + Arrays.toString(sample));
        System.out.println("output: " + Arrays.toString(sort(sample)));
    }
}

File: 29Items-Projects/7Algorithms/1-Sorting-and-Searching/src/python/algorithms/sorting/insertion_sort.py
def insertion_sort(values: list[int]) -> list[int]:
    result = values[:]
    for i in range(1, len(result)):
        key = result[i]
        j = i - 1
        while j >= 0 and result[j] > key:
            result[j + 1] = result[j]
            j -= 1
        result[j + 1] = key
    return result


if __name__ == "__main__":
    sample = [42, 7, 19, 3, 7, 99, 1]
    print("InsertionSort")
    print(f"input : {sample}")
    print(f"output: {insertion_sort(sample)}")
